### PR TITLE
Visitor can have the option to create an article WITH CATEGORIES

### PR DESCRIPTION
**Pull request by boamatule**
# [PT Story](https://www.pivotaltracker.com/story/show/166152109)

## Changes proposed in this pull request:
* Adds Unit Test for the Category model
* Adds Category models
* Adds Associations between Articles and Categories
* Adds Category Controller
* Adds Category to routes
* Adds Seed to populate the Category table

## What I have learned working on this feature:
* Handling Associations

### CODE REVIEW

File: app/controllers/api/v1/category_controller.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+class CategoryController < ApplicationController
```
Comment (raptorf1) [inline app/controllers/api/v1/category_controller.rb]: do we need 2 controllers (`category` and `categories`)????

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,2 +1,14 @@
 class ApplicationController < ActionController::API
```
Comment (raptorf1) [inline app/controllers/application_controller.rb]: why are we putting code in the `ApplicationController`?

File: db/migrate/20190525141132_category_id_to_articles.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class CategoryIdToArticles < ActiveRecord::Migration[5.2]
```
Comment (raptorf1) [inline db/migrate/20190525141132_category_id_to_articles.rb:1]: the same thing is reffered on the first migration file also. any reason for that?

File: spec/requests/api/v1/articles_spec.rb [hunk 4]
```diff
@@ -5,7 +5,8 @@
 
   describe 'GET /api/v1/articles' do
     before do
-      5.times { FactoryBot.create(:article) }
+      category = create(:category)
```
Comment (raptorf1) [inline spec/requests/api/v1/articles_spec.rb]: if the Factory creates articles with categories, do we need that line?

File: app/controllers/api/v1/categories_controller.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+  def create
+    category = Category.create(category_params)
+    if category.persisted?
+      render json: { message: 'Successfully created' }
+    else
+      render status: 422, json: { error: category.errors.full_messages }
+    end
+  end
+
```
Comment (zanenkn) [inline app/controllers/api/v1/categories_controller.rb]: will we be creating new categories? imo unnecessary method.

File: config/routes.rb [hunk 6]
```diff
@@ -4,7 +4,7 @@
       resources :pings, only: [:index], constraints: { format: 'json' }
     end
     namespace :v1 do
-      resources :articles, only: [:index, :create, :show]
+      resources :categories, :articles, only: [:index, :create, :show]
     end
```
Comment (zanenkn) [inline config/routes.rb:9]: id say we should limit the routes open for the `:categories`, for example we wont be editing or deleting categories so those should definitely be excluded.

File: db/seeds.rb [hunk 7]
```diff
@@ -1,7 +1,42 @@
...
+  [ "Science" ],
+  [ "Health" ],
+  [ "Sport" ],
+  [ "Arts" ],
+  [ "Books" ],
+  [ "Style" ],
+  [ "Food" ],
+  [ "Travel" ],
+  [ "Real State" ]
+]
```
Comment (zanenkn) [inline db/seeds.rb:14]: "Real Estate" i guess.

**Review by tochman — CHANGES_REQUESTED** [review]
I can't really see how this constitutes a feature nor how it fulfill the request implied in the PT title/user story. Also, please make sure to name your PR's the same as the corresponding PT story.

File: app/controllers/api/v1/categories_controller.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+class Api::V1::CategoriesController < ApplicationController
+
+  def index
+    categories = Category.all
+    render json: { data: categories }, each_serializer: Categories::IndexSerializer
```
Comment (tochman) [inline app/controllers/api/v1/categories_controller.rb]: Why do you use `data` as a key name? Why not use the resource name?

File: spec/models/article_spec.rb [hunk 9]
```diff
@@ -1,3 +1,5 @@
+require 'rails_helper'
```
Comment (tochman) [inline spec/models/article_spec.rb:1]: Are you sure you need this require? Have you checked the other specs? What are you including in your `.rspec`?

Comment (boamatule) [reply spec/models/article_spec.rb:1]: We saw that we have the `require rails_helper` code  included in the .rspec. 
Removed now.

File: app/controllers/api/v1/categories_controller.rb (hunk 8 above)
Comment (boamatule) [reply app/controllers/api/v1/categories_controller.rb]: We just changed the name to categories which is the right resource name.

File: db/seeds.rb (hunk 7 above)
Comment (oliverochman) [reply db/seeds.rb:14]: please update the "Real State" string

File: spec/models/article_spec.rb [hunk 10]
```diff
@@ -14,11 +17,12 @@
     it { is_expected.to validate_presence_of(:body) }
     it { is_expected.to validate_presence_of(:image) }
     it { is_expected.to validate_presence_of(:written_by) }
+    it { is_expected.to validate_presence_of(:category_id) }
   end
-
+
```
Comment (oliverochman) [inline spec/models/article_spec.rb:26]: You need to test the association between article and category here

**Review by oliverochman — CHANGES_REQUESTED** [review]

**Comment by raptorf1** [comment]
> I can't really see how this constitutes a feature nor how it fulfill the request implied in the PT title/user story. Also, please make sure to name your PR's the same as the corresponding PT story.

User story description changed. It is marked as an API story only. There is another one for the CLIENT side.

**Review by leiter007 — COMMENTED** [review]
Looks great, good adjustments after latest comments.

File: db/seeds.rb (hunk 7 above)
Comment (raptorf1) [reply db/seeds.rb:14]: done

File: spec/models/article_spec.rb (hunk 10 above)
Comment (raptorf1) [reply spec/models/article_spec.rb:26]: done

**Review by oliverochman — APPROVED** [review]

**Pull request merged**